File: binary_search.py
from typing import Optional

def binary_search(array: list[int], target: int) -> Optional[int]:
    start, end = 0, len(array) - 1
    while start <= end:
        mid = (start + end) // 2
        if array[mid] == target:
            return mid
        elif array[mid] < target:
            start = mid + 1
        elif array[mid] > target:
            end = mid - 1
    
    return None

def binary_serach(array: list[int], target: int, start: int, end: int) -> Optional[int]:
    if start > end:
        return None
    
    mid = (start + end) // 2
    if array[mid] == target:
        return mid
    elif array[mid] < target:
        return binary_search(array, target, mid + 1, end)
    elif array[mid] > target:
        return binary_search(array, target, start, mid - 1)
    
    

File: binary search/BinarySearch.java
public class BinarySearch {

    /*
     * Example
     * array = {1, 1, 2, 2, 2, 4, 4, 5, 5, 5}
     * binarySearchLeft(array, 2) = 2
     * binarySearchLeft(array, 3) = 5
     * binarySearchLeft(array, 4) = 5
     * binarySearchRight(array, 2) = 5
     * binarySearchRight(array, 3) = 5
     * binarySearchRight(array, 4) = 7
     */

    public static int binarySearchLeft(int[] array, int target) {

        /*
         * Returns an insertion point (index)
         * which comes before (to the left of) any existing entries of `target` in `array`
         */

        int low = 0, high = array.length;

        while (low < high) {
            int mid = low + (high - low) / 2;
            if (array[mid] < target) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }

        return low;
    }

    public static int binarySearchRight(int[] array, int target) {

        /*
         * Returns an insertion point (index)
         * which comes after (to the right of) any existing entries of `target` in `array`
         */
        
        int low = 0, high = array.length;

        while (low < high) {
            int mid = low + (high - low) / 2;
            if (array[mid] > target) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }

        return low;
    }
}
